bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,963 @@
+	if err != nil {
+		cancel()
+		return nil, stat.Error(codes.Internal,
+			fmt.Sprintf("%s: failed to create client: %v", logLabel, err))
+	}
+
+	if s.clientIDs == nil {
+		s.clientIDs = make(map[string]testClient)
+	}
+
+	s.clientIDs[req.ClientId] = testClient{
+		c: c,
+		cancels: []context.CancelFunc{
+			cancel,
+		},
+		appProfileID:        req.AppProfileId,
+		perOperationTimeout: req.PerOperationTimeout,
+	}
+
+	res := &pb.CreateClientResponse{}
+
+	return res, nil
+}
+
+// RemoveClient responds to the RemoveClient RPC. This method removes an
+// existing client from the goTestProxyServer
+func (s *goTestProxyServer) RemoveClient(ctx context.Context, req *pb.RemoveClientRequest) (*pb.RemoveClientResponse, error) {
+	clientID := req.ClientId
+	doCancelAll := req.CancelAll
+
+	btc, exists := s.clientIDs[clientID]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	if doCancelAll {
+		btc.cancelAll()
+	}
+	btc.c.Close()
+	delete(s.clientIDs, clientID)
+
+	resp := &pb.RemoveClientResponse{}
+	return resp, nil
+}
+
+// ReadRow responds to the ReadRow RPC. This method gets all of the column
+// data for a single row in the Table.
+func (s *goTestProxyServer) ReadRow(ctx context.Context, req *pb.ReadRowRequest) (*pb.RowResult, error) {
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	tName := req.TableName
+	t := btc.c.Open(tName)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	r, err := t.ReadRow(ctx, req.RowKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if r == nil {
+		return nil, fmt.Errorf("%s: no error or row returned from ReadRow()", logLabel)
+	}
+
+	pbRow, err := rowToProto(r)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &pb.RowResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Row: pbRow,
+	}
+
+	return res, nil
+}
+
+// ReadRows responds to the ReadRows RPC. This method gets all of the column
+// data for a set of rows, a range of rows, or the entire table.
+func (s *goTestProxyServer) ReadRows(ctx context.Context, req *pb.ReadRowsRequest) (*pb.RowsResult, error) {
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	lim := req.GetCancelAfterRows()
+
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to ReadRows() is missing inner request")
+	}
+
+	t := btc.c.Open(rrq.TableName)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	rowPbs := rrq.Rows
+	var rs bigtable.RowSet
+
+	// Bigtable client doesn't have a Table.GetAll() function--RowSet must be
+	// provided for ReadRows. Use InfiniteRange() to get the full table.
+	if len(rowPbs.GetRowKeys()) == 0 && len(rowPbs.GetRowRanges()) == 0 {
```

Comment:
Whenever you have an if/else that assigns into the same variable in both branches, consider initializing the variable to a default value, and overriding it under a if statement. The else can be removed in that case.

Reply:
I typically try to do that. However, in this case there's some nasty type casting that I'm trying to work around.